libraries/chain/block_log.cpp

```diff
@@ -327,6 +425,122 @@ namespace eosio { namespace chain {
       index_file.flush();
    }
 
+   size_t detail::block_log_impl::convert_existing_header_to_vacuumed() {
+      uint32_t old_version;
+      uint32_t old_first_block_num;
+      const auto totem = block_log::npos;
+
+      block_file.seek(0);
+      fc::raw::unpack(block_file, old_version);
+      fc::raw::unpack(block_file, old_first_block_num);
+      is_pruned_log_and_mask_version(old_version);
+
+      //we'll always write a v3 log, but need to possibly mutate the genesis_state to a chainid
+      if(first_block_num == 1) {
+         EOS_ASSERT(old_first_block_num == 1, block_log_exception, "expected an old first blocknum of 1");
+         genesis_state gs;
+         auto ds = block_file.create_datastream();
+         fc::raw::unpack(ds, gs);
+
+         block_file.seek(0);
+         fc::raw::pack(block_file, block_log::max_supported_version);
+         fc::raw::pack(block_file, first_block_num);
+         fc::raw::pack(block_file, gs);
+         fc::raw::pack(block_file, totem);
+      }
+      else if(block_log::contains_genesis_state(old_version, old_first_block_num)) {
+         //read in the genesis state and convert it to chainid before writing out
+         genesis_state gs;
+         auto ds = block_file.create_datastream();
+         fc::raw::unpack(ds, gs);
+
+         block_file.seek(0);
+         fc::raw::pack(block_file, block_log::max_supported_version);
+         fc::raw::pack(block_file, first_block_num);
+         fc::raw::pack(block_file, gs.compute_chain_id());
+         fc::raw::pack(block_file, totem);
+      }
+      else {
+         //read in the existing chainid, to parrot back out
+         fc::sha256 chainid;
+         fc::raw::unpack(block_file, chainid);
+
+         block_file.seek(0);
+         fc::raw::pack(block_file, block_log::max_supported_version);
+         fc::raw::pack(block_file, first_block_num);
+         fc::raw::pack(block_file, chainid);
+         fc::raw::pack(block_file, totem);
+      }
+
+      return block_file.tellp();
+   }
+
+   void detail::block_log_impl::vacuum() {
+      //go ahead and write a new valid header now. if the vacuum fails midway, at least this means maybe the
+      // block recovery can get through some blocks.
+      size_t copy_to_pos = convert_existing_header_to_vacuumed();
+
+      version = block_log::max_supported_version;
+      prune_blocks.reset();
+
+      //if there is no head block though, bail now, otherwise first_block_num won't actually be available
+      // and it'll mess this all up
+      if(!head)
```

Comment:
Not following this comment, help me out. No head block means no entries in the log which should mean index is completely empty by my understanding. But digging on this some did show another issue and I’ve added a unit test to address prune transitions on an empty blog.

Reply:
Once someone enables pruning the header will have the flag set that its a pruned log, so when vacuuming (i.e. converting it back to a non-pruned log) it firsts clears the flag by way of rewriting the header via `convert_existing_header_to_vacuumed()`. Then if `!head` then there are no blocks & the index is empty so we're done. Maybe the comment saying "bail" has the wrong connotation -- it's really that we're done here. Otherwise if there are blocks (head is nonnull) the magic has to happen.

Reply:
I agree now with your original code, but I think your new code is wrong. I don't see the convert_existing_header_to_vacuumed() adding the extra "valid non-pruned blocks count" (or whatever that name is) 32 bit number at the end, so why are you removing that 32 bits from the end?

Reply:
(I'm going to refer to the 4 byte block count on the end of the file as the _trailer_)

Creation of a new log is done in `reset()`. This can create a new log initially pruned or nonpruned. It'll write out the header with the correct flag (at the end of the function), as well as write out the trailer either by way of `append()` if there is a first block, or by writing out 0 manually if not (`head==nullptr`).

Conversion from non-pruned -> pruned occurs in `open()`. There, after calling `prune()` it _always_ writes a new header with the flag enabled and the 4 byte trailer to the file. If `head==nullptr` it'll write out a 0.

`vacuum()` is called to transition pruned -> non-pruned. The first thing it does is call `convert_existing_header_to_vacuumed()` which EOS_ASSERTs that we are indeed currently a pruned log file on disk. Then it writes a new header which will have the flag cleared. (it also may need to transition from genesis_state to chain_id). Once the header is written it'll either a) if `head==nullptr` simply chop off the trailer that was added by `reset()`/`open()`, or then if there is a `head` ergo `first_block_num` really exists then it goes on to perform the vacuum procedure which effectively eliminates the trailer.

Reply:
the original code had a defect where if a log was in pruned mode with no blocks and you vacummed it (i.e. converted to non pruned), that `if(!head) return` would leave the trailer on the file after clearing the flag. The log then said it was a nonpruned but the trailer was still on it which of course caused it to be a broken log. Hence adding the code to trim off the trailer there